C7/Map/MapViewCamera.cs

```diff
@@ -0,0 +1,116 @@
+using Godot;
+using C7GameData;
+
+namespace C7.Map {
+
+	// MapViewCamera position and zoom should only be modified through the
+	// following methods:
+	// - scaleZoom
+	// - setZoom
+	// - setPosition
+	// This is because these methods will ensure that MapViewCamera handles
+	// world wrapping on the MapView automatically.
+	public partial class MapViewCamera : Camera2D {
+		private readonly float maxZoom = 3.0f;
+		private readonly float minZoom = 0.2f;
+		public float zoomFactor { get; private set; } = 1.0f;
+		private MapView map;
+		private int wrapEdgeTileMargin = 2; // margin in number of tiles to trigger map wrapping
+		private HorizontalWrapState hwrap = HorizontalWrapState.None;
+
+		public void attachToMapView(MapView map) {
+			this.map = map;
+			wrapEdgeMargin = wrapEdgeTileMargin * map.tileSize.X;
+			map.updateAnimations();
+			checkWorldWrap();
+		}
+
+		public override void _Ready() {
+			base._Ready();
+			scaleZoom(zoomFactor);
+		}
+
+		public void scaleZoom(float factor) {
+			zoomFactor = zoomFactor * factor;
+			zoomFactor = Mathf.Clamp(zoomFactor, minZoom, maxZoom);
+			Zoom = Vector2.One * zoomFactor;
+			checkWorldWrap();
+		}
+
+		public void setZoom(float factor) {
+			zoomFactor = Mathf.Clamp(factor, minZoom, maxZoom);
+			Zoom = Vector2.One * zoomFactor;
+			checkWorldWrap();
+		}
+
+		public void setPosition(Vector2 position) {
+			Position = position;
+			checkWorldWrap();
+		}
+		private float wrapEdgeMargin = 0;
+		float wrapRightEdge {get => map.worldEdgeRight - wrapEdgeMargin; }
+		float wrapLeftEdge {get => map.worldEdgeLeft + wrapEdgeMargin; }
+		private bool enteringRightWrap(Rect2 v) => hwrap != HorizontalWrapState.Right && v.End.X >= wrapRightEdge;
+		private bool enteringLeftWrap(Rect2 v) => hwrap != HorizontalWrapState.Left && v.Position.X <= wrapLeftEdge;
+		private bool atEdgeOfRightWrap(Rect2 v) => hwrap == HorizontalWrapState.Right && v.End.X >= wrapRightEdge + map.pixelWidth;
+		private bool atEdgeOfLeftWrap(Rect2 v) => hwrap == HorizontalWrapState.Left && v.Position.X <= wrapLeftEdge - map.pixelWidth;
+		private HorizontalWrapState currentHwrap(Rect2 v) {
+			return v.Position.X <= wrapLeftEdge ? HorizontalWrapState.Left : (v.End.X >= wrapRightEdge ? HorizontalWrapState.Right : HorizontalWrapState.None);
+		}
+
+		// checkWorldWrap determines if the camera is about to spill over the world map and will:
+		// - move the second "wrap" tilemap to the appropriate edge
+		//   to give the illusion of true wrapping tilemap
+		// - teleport the camera one world-width to the left or right when
+		//   only the "wrap" tilemap is in view
+
+		private void checkWorldWrap() {
```

Comment:
How's this going to work when someone makes a tiny map and zooms all the way out? ie if both left and right edges are visible at the same time

Reply:
Also, just want to go ahead and point out that we should be able to handle horizontal and/or vertical wrap.

Reply:
I'm not sure, but I was thinking instead of something super fancy just limit the minimum possible zoom to never be able to span the entirety of a map. The benefits of a tilemap is that in the average case of normal sized maps we get a huge performance boost. Maybe we can just not load maps that are too small until this edge case is handled more elegantly than limiting zoom